### ISSUE TITLE
`reportPossiblyUnboundVariable` incorrectly reported when `assert_never` used

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#9655]
**Describe the bug**
When using `elif` statements ending in `else: assert_never(x)`, Pyright doesn't recognise that that branch is impossible to hit, and therefore variables don't need to be defined in that branch when evaluating whether variables are potentially unbound.

**Code or Screenshots**

[Pyright playground showing issue](https://pyright-play.net/?pythonVersion=3.12&pythonPlatform=Linux&code=CYUwZgBGD20BQA8BcECWA7ALhAPhAzpgE4CUSAsAFAQ1qSr4aECG6AxiIgDRpZlW1BEAJ4QAvBABEACxAAbOdEkDa8-CArUhNURMkBzWMABGwkMq00iITAFci6ETQDEEAHLRsrCCCJFoRFRBlKCQxsxEiCgY2HiEpJqCqPSM6CzsnAg8MfyWQrpSsgpKKjTyyWip6RzcBMS52jriUobQJmYWgtZ2Dk4Qrh5ejr7%2BgZTBoRDhAF5RvLF1CaV0lUyYrDVZ8w2NBTLyip2qchUMaxuZPPE72nut7ebLahrLgsz46kSYAProIABuvkQJGW3XsjlE-QgAEl2AFrGxMHJRNYAA4BTAgYAQd4QNEYgAK0A%2BqGMyIAquhjNBbOhgAA1CKoZhkkBUIA).

```python
def foo(x: int | str):
    if isinstance(x, int):
        y = "hello"
    else:
        y = "goodbye"
    return y  # Not an error

def bar(x: int | str):
    if isinstance(x, int):
        y = "hello"
    elif isinstance(x, str):
        y = "goodbye"
    return y  # Not an error

def baz(x: int | str):
    if isinstance(x, int):
        y = "hello"
    elif isinstance(x, str):
        y = "goodbye"
    else:
        assert_never(x)
    return y  # Incorrectly reported as reportPossiblyUnboundVariable
```

**VS Code extension or command-line**
Found running Pyright 1.1.376 in PyCharm via LSP. Recreated in playground with Pyright 1.1.389.

**Comment by erictraut** [comment]
Pyright incorporates a technique called [narrowing for implied else](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-implied-else). It works in the case where you have omitted the final `else` in an `if/elif` chain. There is no similar logic for a "present else" that is assumed by the programmer to never be taken. 

Since pyright is working as designed here, I don't consider this a bug. I'll change this to an enhancement request and consider it in that light.

I think this is the first time someone has requested this feature. I'll take a look at how difficult it would be to implement. If it's simple, I'll add support for it now. If it turns out to be difficult or it has a negative performance impact on type analysis, I'll reject the feature request for now and see if there's sufficient signal from pyright users to justify doing the work at some point in the future.

**Comment by Bruntaz** [comment]
Hmm looking at this again, I’m questioning whether I made a mistake when reporting this. When recreating the issue in the playground, I forgot to import `assert_never` and the playground doesn’t seem to complain about function calls that it doesn’t recognise (maybe because the code is unreachable?). 

I know that Pyright does handle returns in the `else` properly and it’s not confused by the variable potentially being undefined. I’d guess that `assert_never` returning `NoReturn` would likely hit the same functionality, and when importing it properly in the playground, Pyright doesn’t complain about `y` being undefined anymore. 

I’ll have a look on Monday at the original case I found this in and try again to recreate the bug with a simple example. I guess it’s possible the original code doesn’t have `assert_never` imported either if Pyright doesn’t complain that it’s undefined…

**Comment by erictraut** [comment]
Yes, good point. If I import `assert_never`, the issue goes away.

**Comment by Bruntaz** [comment]
Yeah can confirm, the original code I found was wrong, and then my recreation forgot to import `assert_never`. Thanks anyway!
